File: src-mobile/ios/Shared/HolidayCalendar.swift
import Foundation

/// Versioned, bundled national calendars. No runtime network or date prediction.
nonisolated struct HolidayCalendar: Sendable {
    struct Day: Sendable {
        let isWorkday: Bool
        let names: [String: String]

        func name(language: String) -> String {
            let normalized = language.replacingOccurrences(of: "_", with: "-")
            return names[normalized] ?? names[String(normalized.prefix(2))]
                ?? names["en"] ?? names.sorted(by: { $0.key < $1.key }).first?.value ?? ""
        }
    }

    private struct Payload: Decodable {
        let schemaVersion: Int
        let datasetVersion: String
        let names: [[String: String]]
        let regions: [String: RegionPayload]
    }

    private struct RegionPayload: Decodable {
        let coveredFromYear: Int
        let coveredThroughYear: Int
        let days: [[Int]]
    }

    private struct Region: Sendable {
        let years: ClosedRange<Int>
        let days: [Int: Day]
    }

    enum LoadError: Error { case invalidDataset }

    static let shared: HolidayCalendar = {
        guard let url = Bundle.main.url(forResource: "HolidayTemplates", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let calendar = try? HolidayCalendar(data: data)
        else { return HolidayCalendar() }
        return calendar
    }()

    let datasetVersion: String
    private let regions: [String: Region]
    var regionIdentifiers: [String] { regions.keys.sorted() }

    private init() {
        datasetVersion = ""
        regions = [:]
    }

    init(data: Data) throws {
        let payload = try JSONDecoder().decode(Payload.self, from: data)
        guard payload.schemaVersion == 1, !payload.datasetVersion.isEmpty else {
            throw LoadError.invalidDataset
        }
        var parsed: [String: Region] = [:]
        for (identifier, source) in payload.regions {
            guard Self.isValidRegionIdentifier(identifier), !identifier.isEmpty,
                  source.coveredFromYear <= source.coveredThroughYear else { throw LoadError.invalidDataset }
            var days: [Int: Day] = [:]
            for row in source.days {
                guard row.count == 3, (0...1).contains(row[1]),
                      payload.names.indices.contains(row[2]), days[row[0]] == nil,
                      Self.isValidDateCode(row[0]),
                      (source.coveredFromYear...source.coveredThroughYear).contains(row[0] / 10_000)
                else { throw LoadError.invalidDataset }
                days[row[0]] = Day(isWorkday: row[1] == 1, names: payload.names[row[2]])
            }
            parsed[identifier] = Region(years: source.coveredFromYear...source.coveredThroughYear, days: days)
        }
        datasetVersion = payload.datasetVersion
        regions = parsed
    }

    static func isValidRegionIdentifier(_ identifier: String?) -> Bool {
        guard let identifier, !identifier.isEmpty else { return true }
        return identifier.utf8.count == 2 && identifier.utf8.allSatisfy { (65...90).contains($0) }
    }

    static func isValidDateCode(_ value: Int) -> Bool {
        let year = value / 10_000
        let month = value / 100 % 100
        let day = value % 100
        guard (1...9_999).contains(year), (1...12).contains(month), (1...31).contains(day) else { return false }
        let civil = CivilZone.civilDate(dayNumber: CivilZone.dayNumber(year: year, month: month, day: day))
        return civil.year == year && civil.month == month && civil.day == day
    }

    func defaultRegionIdentifier(locale: Locale = .current) -> String? {
        guard let code = locale.region?.identifier, regions[code] != nil else { return nil }
        return code
    }

    func regionName(_ identifier: String, locale: Locale) -> String {
        // Foundation's region aliases can vary with device configuration even
        // for an explicit locale. Keep the Traditional Chinese display name
        // tied to the app language, including zh-TW and zh-HK overrides.
        if identifier == "TW", locale.language.languageCode?.identifier == "zh",
           locale.language.script?.identifier == "Hant" {
            return "台灣"
        }
        return locale.localizedString(forRegionCode: identifier) ?? identifier
    }

    func coveredThroughYear(regionIdentifier: String) -> Int? {
        regions[regionIdentifier]?.years.upperBound
    }

    /// Transport effects, not names, so paired devices use the same source
    /// revision even when their installed bundles differ.
    func workdayOverrides(regionIdentifier: String) -> [Int: Bool] {
        regions[regionIdentifier]?.days.mapValues(\.isWorkday) ?? [:]
    }

    func covers(year: Int, regionIdentifier: String) -> Bool {
        regions[regionIdentifier]?.years.contains(year) == true
    }

    func day(dayKey: String, regionIdentifier: String) -> Day? {
        guard let parts = ExtendedScheduleResolver.parse(dayKey: dayKey) else { return nil }
        return regions[regionIdentifier]?.days[parts.year * 10_000 + parts.month * 100 + parts.day]
    }

    func day(dayNumber: Int, regionIdentifier: String) -> Day? {
        let civil = CivilZone.civilDate(dayNumber: dayNumber)
        return regions[regionIdentifier]?.days[civil.year * 10_000 + civil.month * 100 + civil.day]
    }
}

File: src-mobile/android/core/domain/src/main/kotlin/com/rainif/doneat/core/domain/schedule/HolidayCalendar.kt
package com.rainif.doneat.core.domain.schedule

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Versioned, bundled national calendars (`HolidayTemplates.json`, shared with
 * iOS). No runtime network or date prediction: outside a region's covered
 * years there is simply no entry. Loaded once by the caller and passed in; the
 * rules never read a global.
 */
class HolidayCalendar private constructor(
    val datasetVersion: String,
    private val regions: Map<String, Region>,
) {
    data class Day(val isWorkday: Boolean, val names: Map<String, String>)

    private class Region(val years: IntRange, val days: Map<Int, Day>)

    val regionIdentifiers: List<String> get() = regions.keys.sorted()

    fun coveredThroughYear(regionIdentifier: String): Int? = regions[regionIdentifier]?.years?.last

    fun covers(year: Int, regionIdentifier: String) = regions[regionIdentifier]?.years?.contains(year) == true

    /** `dateCode` is `yyyymmdd`. */
    fun day(dateCode: Int, regionIdentifier: String): Day? = regions[regionIdentifier]?.days?.get(dateCode)

    /** Effects only, for transport to paired devices whose bundle may differ. */
    fun workdayOverrides(regionIdentifier: String): Map<Int, Boolean> =
        regions[regionIdentifier]?.days?.mapValues { it.value.isWorkday } ?: emptyMap()

    class InvalidDataset(message: String) : IllegalArgumentException(message)

    companion object {
        val EMPTY = HolidayCalendar("", emptyMap())

        /** Parses and validates the whole dataset; any malformed row rejects it. */
        fun parse(json: String): HolidayCalendar {
            val root = Json.parseToJsonElement(json).jsonObject
            fun fail(reason: String): Nothing = throw InvalidDataset(reason)
            if (root["schemaVersion"]?.jsonPrimitive?.int != 1) fail("schemaVersion")
            val version = root["datasetVersion"]?.jsonPrimitive?.content.orEmpty()
            if (version.isEmpty()) fail("datasetVersion")
            val names = root.getValue("names").jsonArray.map { entry ->
                entry.jsonObject.mapValues { it.value.jsonPrimitive.content }
            }
            val regions = root.getValue("regions").jsonObject.mapValues { (identifier, element) ->
                val source = element.jsonObject
                val from = source.getValue("coveredFromYear").jsonPrimitive.int
                val through = source.getValue("coveredThroughYear").jsonPrimitive.int
                if (identifier.isEmpty() || !isValidRegionIdentifier(identifier) || from > through) fail("region $identifier")
                val days = HashMap<Int, Day>()
                for (row in source.getValue("days").jsonArray) {
                    val values = row.jsonArray.map { it.jsonPrimitive.int }
                    if (values.size != 3 || values[1] !in 0..1 || values[2] !in names.indices ||
                        days.containsKey(values[0]) || !isValidDateCode(values[0]) || values[0] / 10_000 !in from..through
                    ) fail("row $identifier $values")
                    days[values[0]] = Day(values[1] == 1, names[values[2]])
                }
                Region(from..through, days)
            }
            return HolidayCalendar(version, regions)
        }

        /** Empty or absent means "no region"; otherwise two uppercase ASCII letters. */
        fun isValidRegionIdentifier(identifier: String?): Boolean {
            if (identifier.isNullOrEmpty()) return true
            return identifier.length == 2 && identifier.all { it in 'A'..'Z' }
        }

        fun isValidDateCode(value: Int): Boolean {
            val year = value / 10_000
            val month = value / 100 % 100
            val day = value % 100
            if (year !in 1..9_999 || month !in 1..12 || day !in 1..31) return false
            return CivilZone.civilDate(CivilZone.dayNumber(year, month, day)) == Triple(year, month, day)
        }
    }
}
